vllm/v1/core/guided_decoding/grammar.py

```diff
@@ -0,0 +1,154 @@
+from typing import (TYPE_CHECKING, Any, Generic, Literal, TypeVar, get_args,
+                    overload)
+
+from typing_extensions import Annotated, LiteralString
+
+from vllm.utils import LazyLoader
+
+if TYPE_CHECKING:
+    import torch
+    import xgrammar as xgr
+    from typing_extensions import Self
+else:
+    xgr = LazyLoader("xgr", globals(), "xgrammar")
+    torch = LazyLoader("torch", globals(), "torch")
+
+T = TypeVar("T", bound=Annotated[LiteralString, str])
+
+
+class Grammar(ABC, Generic[T]):
+    finished: bool = False
+
+    @abstractmethod
+    def accept_token(self, token: int) -> bool:
+        """Whether to accept the token and advance the machine state."""
+
+    @abstractmethod
+    def fill_bitmask(self, bitmask: torch.Tensor, idx: int) -> None:
+        """Fill the bitmask for the token at the given index."""
+
+    @abstractmethod
+    def allocate_bitmask(self, batch_size: int,
+                         vocab_size: int) -> torch.Tensor:
+        """Allocate a bitmask for the given batch size and vocabulary size."""
+
+    @staticmethod
+    @abstractmethod
+    def apply_bitmask(logits: torch.Tensor, vocab_mask: torch.Tensor) -> None:
+        """Apply the bitmask to the logits."""
+
+    @abstractmethod
+    def reset(self):
+        """Reset the machine state."""
+
+    @abstractmethod
+    def copy(self) -> Self:
+        """Copy the grammar object."""
+
+    def __copy__(self):
+        return self.copy()
+
+    _registry: dict[str, type[Grammar[T]]] = {}
+    _backend: T
+
+    def __init_subclass__(cls):
+        if not hasattr(cls, '__orig_bases__'):
+            raise TypeError(
+                f"Class {cls.__qualname__} must be a subclass of GrammarObject"
+            )
+
+        backend = None
+        for base in cls.__orig_bases__:
+            if (origin := get_args(base)) and issubclass(
+                    base.__origin__, Grammar):
+                backend = get_args(origin[0])[0]
+                break
+
+        if backend is None:
+            raise TypeError(
+                f"Class {cls.__qualname__} must specify backend as Literal type"
+            )
+
+        if backend in cls._registry:
+            name = cls._registry[backend].__qualname__
+            raise ValueError(
+                f"Backend '{backend}' is already registered to {name}")
+
+        # Set the backend value from the Literal type
+        cls._backend = backend
+        cls._registry[backend] = cls
+
+    @overload
+    @classmethod
+    def from_backend(
+        cls,
+        backend: Literal["xgrammar"] = ...,
+        *,
+        matcher: xgr.GrammarMatcher = ...,
+        vocab_size: int = ...,
+        ctx: xgr.CompiledGrammar = ...,
+    ) -> XGrammar:
+        ...
+
+    @overload
+    @classmethod
+    def from_backend(
+        cls,
+        backend: Literal["outlines"] = ...,
+        *,
+        guide: str = ...,
+        whitespace_pattern: str | None = ...,
+    ) -> XGrammar:
```

Comment:
This doesn't look right -- `Literal["outlines"]`  -- `-> XGrammar`

Reply:
Oh, you are right, I was just putting this as a placeholder for now, but will update this promptly.